### PR TITLE
[Microsoft.App] targetLabel and Labels revision mode

### PR DESCRIPTION
**Pull request by smzeng**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

spec_pr_review_workflow_diagram

## Purpose of this PR

add new property targetLabel and also new activeRevisionsMode labels
original PR here, but closed due to merge issues https://github.com/Azure/azure-rest-api-specs/pull/30509

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [ ] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [ ] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

### CODE REVIEW

**Review by Tratcher — APPROVED** [review]

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    "https"
+  ],
+  "consumes": [
+    "application/json"
+  ],
+  "produces": [
+    "application/json"
+  ],
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/containerApps/{containerAppName}/labelHistory": {
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json:18]: Make sure PMs have signed off on new `labelHistory` resource type name

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  },
+  "responses": {
+    "200": {
+      "headers": {},
+      "body": {
+          "id": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/rg/providers/Microsoft.App/containerApps/testcontainerApp0/labelHistory/dev",
+          "name": "dev",
+          "type": "Microsoft.App/containerApps/labelHistory",
+          "properties": {
+            "label": "dev",
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json:17]: if `name` property already has label name, this seems redundant

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    "api-version": "2024-10-02-preview"
+  },
+  "responses": {
+    "200": {
+      "headers": {},
+      "body": {
+          "id": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/rg/providers/Microsoft.App/containerApps/testcontainerApp0/labelHistory/dev",
+          "name": "dev",
+          "type": "Microsoft.App/containerApps/labelHistory",
+          "properties": {
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json:16]: I assumed history would be an array of changes

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_List.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        "value": [
+          {
+            "id": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/rg/providers/Microsoft.App/containerApps/testcontainerApp0/labelHistory/prod",
+            "name": "prod",
+            "type": "Microsoft.App/containerApps/labelHistory",
+            "properties": {
+              "label": "prod",
+              "revision": "testcontainerApp0-pjxhsye",
+              "start": "2024-09-24T21:20:22+00:00",
+              "stop": "2024-09-24T21:24:22+00:00"
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_List.json:21]: Thought - Will the date time property have time zone offset? Do you have to keep it UTC always?

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerApps.json [hunk 5]
```diff
@@ -640,6 +641,10 @@
             "modelAsString": true
           }
         },
+        "targetLabel": {
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerApps.json:644]: Update existing example for this property

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+      "get": {
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json:19]: Are there "put" and "patch" operations?

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsRevisions.json [hunk 7]
```diff
@@ -442,6 +442,14 @@
               "type": "integer",
               "readOnly": true
             },
+            "labels": {
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsRevisions.json:445]: Update existing examepl

**Review by jijohn14 — CHANGES_REQUESTED** [review]

**Comment by rkmanda** [comment]
Pl fix the errors here - https://github.com/Azure/azure-rest-api-specs/pull/30802/checks?check_run_id=30881064450

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json (hunk 1 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json:18]: we went over label history with Elliot, Korede, and Anthony :)

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_List.json (hunk 4 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_List.json:21]: In the code we are keeping it UTC right so it is uniform
Time zone is a good idea for future improvement though

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json (hunk 6 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json:19]: label history only has GET, to put or patch you have to use the container apps traffic section

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerApps.json (hunk 5 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerApps.json:644]: addded in latest commit

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json (hunk 2 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json]: the name here should have been the containerapp name fixed it

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json (hunk 3 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json]: thanks for catching this! fixed, asking @Tratcher to do another review

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_List.json [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    "containerAppName": "labelsapp",
+    "api-version": "2024-10-02-preview"
+  },
+  "responses": {
+    "200": {
+      "headers": {},
+      "body": {
+        "value": [
+          {
+            "id": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/rg/providers/Microsoft.App/containerApps/labelsapp",
```
Comment (Tratcher) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_List.json:14]:
```suggestion
            "id": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/rg/providers/Microsoft.App/containerApps/labelsapp/labelHistory",
```

Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_List.json]: fixed based on our convo

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    "LabelHistory": {
+      "description": "Container App Label History.",
+      "type": "object",
+      "allOf": [
+        {
+          "$ref": "../../../../../common-types/resource-management/v5/types.json#/definitions/ProxyResource"
+        }
+      ],
+      "properties": {
+        "properties": {
```
Comment (Tratcher) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json:142]: properties/properties?

**Review by Tratcher — APPROVED** [review]

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        "operationId": "ContainerAppsLabelHistory_ListLabelHistory",
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "name": "containerAppName",
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json:33]: Could you move this property to CommonDefinitions? This is reference in many swagger definitions.

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+      "properties": {
+        "revision": {
+          "description": "Container App revision name that label was applied to.",
+          "type": "string",
+          "readOnly": true
+        },
+        "start": {
+          "format": "date-time",
+          "description": "Timestamp describing when the label was applied to the revision.",
+          "type": "string",
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json:174]: Should it be DateTime?

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/ContainerApps_CreateOrUpdate.json [hunk 12]
```diff
@@ -16,6 +16,7 @@
         "environmentId": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/rg/providers/Microsoft.App/managedEnvironments/demokube",
         "workloadProfileName": "My-GP-01",
         "configuration": {
+          "targetLabel": "production",
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/ContainerApps_CreateOrUpdate.json:19]: i think we have a separate example file for labels and keep default as is

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json [hunk 13]
```diff
@@ -0,0 +1,32 @@
+{
+  "parameters": {
+    "subscriptionId": "34adfa4f-cedf-4dc0-ba29-b6d1a69ab345",
+    "resourceGroupName": "rg",
+    "containerAppName": "labelsapp",
```
Comment (jijohn14) [inline specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json:5]: nit: can ignore. To read label properties separately

```suggestion
    "containerAppName": "testContainerApp",
```

**Review by jijohn14 — APPROVED** [review]

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json (hunk 11 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json:164]: I think date-time is correct https://swagger.io/docs/specification/v3_0/data-models/data-types/

i think DateTime a datatype used for tables

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json (hunk 13 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/LabelHistory_Get.json]: good point, for readability i fixed it

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/ContainerApps_CreateOrUpdate.json (hunk 12 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/examples/ContainerApps_CreateOrUpdate.json]: i agree, fixed

File: specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json (hunk 10 above)
Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json]: moved :)

Comment (smzeng) [reply specification/app/resource-manager/Microsoft.App/preview/2024-10-02-preview/ContainerAppsLabelHistory.json]: I think theres too many places using it rn to move to ref the common defn in this PR. I can volunteer to clean up the code and have them ref the new common defn parameter in the future

**Comment by smzeng** [comment]
@rkmanda all the issues have been addressed this is ready for review again

**Pull request merged**